New Incremental Search Panel Inside The results Window

Calisto-Mathias opened the pull request:
This commit has all the work that I have done on the incremental search panel. I am yet to make the code in a better way by using better practices but I intend to do that over the next few days. At the moment, my buildtools on a clean repo is still building. As soon as this is over, I will refactor the important parts of this in an attempt to do better memory management, etc.

I have also tested this with the lib_debug.so as stated in the email and fixed the issue that was causing the frequent crashes.

There is still some work to be done so I might need an extension. I also need to refactor a bit of this but my buildtools is suddenly taking way too long to compile and was popping up errors and im recompiling them at the moment.

Calisto-Mathias committed:
New Incremental Search Panel Inside The results Window

This commit has all the work that I have done on the incremental search panel. I am yet to make the code in a better way by using better practices but I intend to do that over the next few days. At the moment, my buildtools on a clean repo is still building. As soon as this is over, I will refactor the important parts of this in an attempt to do better memory management, etc.

I have also tested this with the lib_debug.so as stated in the email and fixed the issue that was causing the frequent crashes.

Calisto-Mathias committed:
Incremental Find Panel - Refactor

I made sure to use more safe ways of writing the code in this commit. This includes working as much as possible with C++ objects that follow RAII versus just using raw C-style objects.

Apart from this, I also made a few small changes such as rewriting most functions to return a status_t value so that I could use these return values in the future for better checking of errors to increase robustness.

A few things are still left that need to be done:
	1) Resizing the entire Incremental Search Panel according to the columns present. This includes setting a minimum size based on the columns.
	2) Adding the ability to combine the columns using the "OR" combinator
	3) Trimming down the code in a few sections to make it shorter and cleaner

Apart from these short-term goals, some long-term goals with this panel are:
	1) Find a way to debounce queries so that the button can be discarded and searches can be executed simply by typing. I did explore this avenue but it seems to be bombarding the back-end with too many messages at the moment.
	2) Find a way to store and restore the columns accurately so that even the incremental queries can be saved :)

nielx commented on src/kits/tracker/PopUpTextControl.h:
```diff
@@ -0,0 +1,36 @@
+	virtual	void				MouseDown(BPoint) override;
+private:
+			BPopUpMenu*			fOptionsMenu;
+			BHandler*			fTarget;
```
If the only goal for this member is to send messages, I suggest you create a BMessenger, rather than creating it in the MessageReceived() hook.

nielx commented on src/kits/tracker/PoseView.cpp:
```diff
@@ -35,6 +35,7 @@ All rights reserved.
 
 #include "PoseView.h"
 
+#include <iostream>
```
What are you using form iostream?

Calisto-Mathias replied:
I removed all the debug output and I seem to have forgotten to remove the header file from being included. I fixed this in the latest commit.

nielx commented on src/kits/tracker/QueryPoseView.cpp:
```diff
@@ -144,6 +174,26 @@ BQueryPoseView::MessageReceived(BMessage* message)
+				break;
+			
+			delete fModel;
+			fModel = new Model(&ref);
```
Do you also need to delete the model in the destructor?

nielx commented on src/kits/tracker/TFindPanelLooper.cpp:
```diff
@@ -0,0 +1,59 @@
+				// std::cout<<"Sent Message"<<std::endl;
+			} else {
+				BMessageQueue* queue = MessageQueue();
+				std::cout<<"Length:"<<queue->CountMessages()<<std::endl;
```
Please remove debug output.

Calisto-Mathias replied:
Done in the latest commit. I think I forgot to take them out :(

nielx commented on src/kits/tracker/TitleView.cpp:
```diff
@@ -45,6 +45,7 @@ All rights reserved.
 #include <PopUpMenu.h>
 #include <Window.h>
 
+#include <iostream>
```
Unused?

Calisto-Mathias replied:
Yup. I removed this in the latest commit.

nielx reviewed (commented):
There might be more detailed commentary to follow, but there is nothing obviously bad practise in the PR, so I think we should move the discussion to Gerrit to have more eyes on it.

Calisto-Mathias committed:
Add Scrolling Synchronisation For Column Container

This PR adds in the ability for the fColumnsContainer to smoothly scroll along with the PoseView below it. This is crucial from a user point of view.

Calisto-Mathias commented:
> There might be more detailed commentary to follow, but there is nothing obviously bad practise in the PR, so I think we should move the discussion to Gerrit to have more eyes on it.

Yup. I think we should move it to Gerrit as well. I'll try to get the last task completed (giving a minimum size to the columns) as soon as possible and push a commit here. If everything checks out, I'll move it to Gerrit in the next 1-3 days.

Apart from this, I did implement synchronisation between the scrolling of the PoseView and the Columns Container.

Calisto-Mathias committed:
Fix minimum sizing for columns

In this commit, I have fixed the issue being faced by columns that have a default size that is lesser than the value of kMinColumnWidth.

With this, I believe that the first version of the incremental search panel is complete :)

Calisto-Mathias commented:
In the latest commit, I have fixed some small issues. I think with this, the first version of the incremental search panel can successfully be integrated into the Find panel Results window. It also has the same search speed. We may need to re-index the files though (as I have noticed that after the initial results are cached, the load times are much faster). I asked PulkoMandy about this and he said as of now the caching system works in that way. 

I will try and shorten down the code (Maybe clean up some parts of it). I'll put it up as an MR on Gerrit once that is done.
